### PR TITLE
Do step 12. add the Order and OrderItem entity project

### PR DESCRIPTION
**Pull request by yurlis**
…eeded controller's endpoints.

### CODE REVIEW

File: src/main/java/com/bookstoreapp/controller/OrderController.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        return ResponseEntity.ok().build();
+    }
+
+    @GetMapping("/{orderId}/items")
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @Operation(summary = "Get all order items",
+            description = "Get all items for the specific order")
+    List<OrderItemDto> getAllOrderItems(Authentication authentication,
+                                        @PathVariable Long orderId) {
+        User user = (User) authentication.getPrincipal();
```
Comment (MAMentorFX) [inline src/main/java/com/bookstoreapp/controller/OrderController.java]: extract to local method

File: src/main/java/com/bookstoreapp/mapper/OrderItemMapper.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+import org.mapstruct.Mapping;
+
+@Mapper(config = MapperConfig.class)
+public interface OrderItemMapper {
+    @Mapping(target = "price", ignore = true)
+    @Mapping(target = "order", ignore = true)
+    OrderItem toOrderItemModel(CartItem cartItem);
+
+    @Mapping(target = "bookId", source = "book.id")
+    @Mapping(target = "quantity", source = "quantity")
```
Comment (MAMentorFX) [inline src/main/java/com/bookstoreapp/mapper/OrderItemMapper.java]:
```suggestion
```

File: src/main/java/com/bookstoreapp/mapper/OrderMapper.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+
+@Mapper(config = MapperConfig.class, uses = OrderItemMapper.class)
+public interface OrderMapper {
+    @Mapping(target = "userId", source = "user.id")
+    @Mapping(target = "orderItems", source = "orderItems")
+    @Mapping(target = "orderDate", source = "orderDate")
+    @Mapping(target = "total", source = "total")
+    @Mapping(target = "status", source = "status")
```
Comment (MAMentorFX) [inline src/main/java/com/bookstoreapp/mapper/OrderMapper.java]:
```suggestion
```

File: src/main/java/com/bookstoreapp/service/OrderServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+import java.time.LocalDateTime;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@Service
+@RequiredArgsConstructor
+public class OrderServiceImpl implements OrderService {
+    @PersistenceContext
+    private EntityManager entityManager;
```
Comment (MAMentorFX) [inline src/main/java/com/bookstoreapp/service/OrderServiceImpl.java]: tell me the reason for this field here? I think it's redundant

File: src/main/java/com/bookstoreapp/service/OrderServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+                .map(orderMapper::toDto)
+                .toList();
+    }
+
+    @Override
+    public List<OrderItemDto> findAllOrderItemsByOrderId(Long userId, Long orderId) {
+        Order orderFromDb = getOrderFromDbByUserIdAndOrderId(userId, orderId);
+        return orderFromDb.getOrderItems().stream()
+                .map(orderItemMapper::toDto)
+                .toList();
```
Comment (MAMentorFX) [inline src/main/java/com/bookstoreapp/service/OrderServiceImpl.java]: Mapstruct supports mapping collections from the box

File: src/main/java/com/bookstoreapp/service/OrderServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+        cartItemRepository.deleteAll(cartFromDB.getCartItems());
+        cartFromDB.getCartItems().clear();
+        return orderMapper.toDto(order);
+    }
+
+    @Override
+    public List<OrderDto> findAll(Long userId, Pageable pageable) {
+        return orderRepository.findAllByUserId(userId, pageable).stream()
+                .map(orderMapper::toDto)
+                .toList();
```
Comment (MAMentorFX) [inline src/main/java/com/bookstoreapp/service/OrderServiceImpl.java]: Mapstruct supports mapping collections from the box

**Review by MAMentorFX — CHANGES_REQUESTED** [review]

File: src/main/java/com/bookstoreapp/service/OrderServiceImpl.java (hunk 4 above)
Comment (yurlis) [reply src/main/java/com/bookstoreapp/service/OrderServiceImpl.java]: The search for a solution was difficult. We couldn't find a way to bind order_items. I forgot to delete it.

File: src/main/java/com/bookstoreapp/service/OrderServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+    }
+
+    @Override
+    public List<OrderItemDto> findAllOrderItemsByOrderId(Long userId, Long orderId) {
+        Order orderFromDb = getOrderFromDbByUserIdAndOrderId(userId, orderId);
+        Set<OrderItem> orderItems = orderFromDb.getOrderItems();
+        return orderItemMapper.toDto(orderItems);
+//        return orderFromDb.getOrderItems().stream()
+//                .map(orderItemMapper::toDto)
+//                .toList();
```
Comment (okuzan) [inline src/main/java/com/bookstoreapp/service/OrderServiceImpl.java]: don't forget to remove commented out code

File: src/main/java/com/bookstoreapp/model/Order.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.SQLRestriction;
+
+import java.math.BigDecimal;
+import java.time.LocalDateTime;
+import java.util.Set;
+
+@Entity
+@SQLRestriction("is_deleted=false")
+@SQLDelete(sql = "UPDATE books SET is_deleted = true WHERE id=?")
```
Comment (okuzan) [inline src/main/java/com/bookstoreapp/model/Order.java]:
```suggestion
@SQLRestriction("is_deleted = FALSE")
@SQLDelete(sql = "UPDATE books SET is_deleted = TRUE WHERE id = ?")
```

File: src/main/java/com/bookstoreapp/service/OrderServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+        final Order finalOrder = order;
+        Set<OrderItem> orderItems = cartFromDB.getCartItems().stream()
+                .map(cartItem -> {
+                    OrderItem orderItem = orderItemMapper.toOrderItemModel(cartItem);
+                    orderItem.setOrder(finalOrder);
+                    Book book = cartItem.getBook();
+                    orderItem.setPrice(book.getPrice());
+                    orderItem.setQuantity(cartItem.getQuantity());
+                    return orderItem;
+                })
```
Comment (okuzan) [inline src/main/java/com/bookstoreapp/service/OrderServiceImpl.java]: try to extract to a private method for clarity

File: src/main/java/com/bookstoreapp/service/OrderServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+    public OrderDto save(Long userId, PlaceOrderRequestDto requestDto) {
+        Order order = new Order();
+        ShoppingCart cartFromDB = getCartFromDB(userId);
+
+        order.setUser(userRepository.findById(userId).orElseThrow(
+                () -> new RuntimeException("Cannot find user by user id: " + userId)));
+        order.setShippingAddress(requestDto.shippingAddress());
+        order.setStatus(Order.Status.PENDING);
+        order.setOrderDate(LocalDateTime.now());
+        order.setTotal(BigDecimal.valueOf(0));
```
Comment (okuzan) [inline src/main/java/com/bookstoreapp/service/OrderServiceImpl.java]:
```suggestion
        order.setTotal(BigDecimal.ZERO);
```

File: src/main/java/com/bookstoreapp/service/OrderServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+    private final UserRepository userRepository;
+
+    @Override
+    @Transactional
+    public OrderDto save(Long userId, PlaceOrderRequestDto requestDto) {
+        Order order = new Order();
+        ShoppingCart cartFromDB = getCartFromDB(userId);
+
+        order.setUser(userRepository.findById(userId).orElseThrow(
+                () -> new RuntimeException("Cannot find user by user id: " + userId)));
```
Comment (okuzan) [inline src/main/java/com/bookstoreapp/service/OrderServiceImpl.java]: EntityNotFound

File: src/main/java/com/bookstoreapp/service/OrderServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+        return orderMapper.toDto(order);
+    }
+
+    @Override
+    public List<OrderDto> findAll(Long userId, Pageable pageable) {
+        List<Order> orders = orderRepository.findAllByUserId(userId, pageable);
+        return orderMapper.toDto(orders);
+//        return orderRepository.findAllByUserId(userId, pageable).stream()
+//                .map(orderMapper::toDto)
+//                .toList();
```
Comment (okuzan) [inline src/main/java/com/bookstoreapp/service/OrderServiceImpl.java]: same

**Review by okuzan — CHANGES_REQUESTED** [review]

**Review by sarakhmen — APPROVED** [review]

**Pull request merged**